### PR TITLE
feat(application): feat add sdk

### PR DESCRIPTION
**Pull request by thewbuk**
Because

- application sdk

### CODE REVIEW

File: packages/sdk/src/application/ApplicationClient.ts
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:1]: We miss these endpoints

- https://openapi.instill.tech/reference/apppublicservice_listapps
- https://openapi.instill.tech/reference/apppublicservice_deleteconversation
- ...

@thewbuk I am not going to do a 1-1 check for you, but please include all the endpoint we list in the API reference. We need all of them

Every request has it own response type, for example createConversation should have the type as 

```
type CreateConversationResponse = {
  conversation: Conversation
}
```

File: packages/sdk/src/application/ApplicationClient.ts [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+      const additionalHeaders = getInstillAdditionalHeaders({});
+      const response = await this._client.get<{ catalogs: Catalog[] }>(
+        `/namespaces/${ownerId}/catalogs`,
+        { additionalHeaders },
+      );
+      return Promise.resolve(response.catalogs || []);
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:244]: @thewbuk This should be in the CatalogClient not ApplicationClient

File: packages/sdk/src/application/ApplicationClient.ts [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+  async listMessages({ ownerId, appId, conversationId }: ListMessagesRequest) {
+    try {
+      const additionalHeaders = getInstillAdditionalHeaders({});
+      const response = await this._client.get<{ messages: Message[] }>(
+        `/namespaces/${ownerId}/apps/${appId}/conversations/${conversationId}/messages?ifAll=true`,
+        { additionalHeaders },
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:186]: @thewbuk User should be able to freely choose whether they want to set up this query parameter, we should make it as parameter

**Review by EiffelFly — COMMENTED** [review]
@thewbuk You successfully grasp the concept of our SDK but there are still lots of missing parts

- Where is CatalogClient -> You include some of the SDK function for catalog in the ApplicationClient, this is confusing
- You are lacking the Response type of every function, they need to be exported
- You miss many endpoint in the ApplicationClient too. Please 1-1 include them in the SDK
- Let's use `getQueryString` functions for all the SDK functions, other SDk should follow this rule and I will refactor them

And there are others small issues around this PR, please check comment

File: packages/sdk/src/application/ApplicationClient.ts [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+            ownerId,
+          })) as Application[]),
+        );
+      }
+
+      return Promise.resolve(applications);
+    } catch (err) {
+      return Promise.reject(err);
+    }
+  }
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:81]: @thewbuk This is totally wrong. These has no overhead, how can user of SDK get the concrete type

File: packages/sdk/src/application/ApplicationClient.ts [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+      const queryString = getQueryString({
+        baseURL: `/v1alpha/namespaces/${ownerId}/apps`,
+        pageSize: pageSize ?? undefined,
+        pageToken: pageToken ?? undefined,
+        view: view ?? undefined,
+      });
+
+      const data = await this._client.get<ListApplicationsResponse>(
+        queryString,
+        {},
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:56]: @thewbuk Why this is a empty object?

File: packages/sdk/src/application/ApplicationClient.ts [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    },
+  ): Promise<ListApplicationsResponse | Application[]> {
+    const { pageSize, pageToken, view, enablePagination, ownerId } = props;
+
+    try {
+      const queryString = getQueryString({
+        baseURL: `/v1alpha/namespaces/${ownerId}/apps`,
+        pageSize: pageSize ?? undefined,
+        pageToken: pageToken ?? undefined,
+        view: view ?? undefined,
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:51]: @thewbuk you can set the parameter as undefined so you don't need to shortcut it here

File: packages/sdk/src/application/ApplicationClient.ts [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+  async deleteApplication({
+    ownerId,
+    appId,
+  }: DeleteApplicationRequest): Promise<void> {
+    try {
+      const queryString = getQueryString({
+        baseURL: `/v1alpha/namespaces/${ownerId}/apps/${appId}`,
+      });
+
+      await this._client.delete(queryString, {});
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:157]: @thewbuk Please remove empty object

File: packages/sdk/src/application/ApplicationClient.ts [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+      const data = await this._client.get<ListConversationsResponse>(
+        queryString,
+        {},
+      );
+
+      return Promise.resolve(data);
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:185]: @thewbuk This should enable pagination

File: packages/sdk/src/application/ApplicationClient.ts [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+
+      const response = await this._client.get<ListMessagesResponse>(
+        queryString,
+        {},
+      );
+      return Promise.resolve(response);
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:264]: @thewbuk listMessages support pageToken and pageSize in the API, so this endpoint should be like all the others list sdk functions, support pagination

File: packages/sdk/src/application/ApplicationClient.ts [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+          queryString,
+          {
+            body: JSON.stringify(payload),
+          },
+        );
+      return Promise.resolve(response);
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
```
Comment (EiffelFly) [inline packages/sdk/src/application/ApplicationClient.ts:388]: @thewbuk This endpoint is not there is the API docs, is BE missed this one?

File: packages/sdk/src/catalog/index.ts
Comment (EiffelFly) [inline packages/sdk/src/catalog/index.ts:1]: If this file doesn't have any meaning, please remove them

File: packages/sdk/src/catalog/CatalogClient.ts
Comment (EiffelFly) [inline packages/sdk/src/catalog/CatalogClient.ts:1]: @thewbuk Please add CatalogClient in this sprint

Comment (EiffelFly) [reply packages/sdk/src/catalog/CatalogClient.ts:1]: I will temp approve this change

**Review by EiffelFly — COMMENTED** [review]
@thewbuk There are two main issues 

- You have many unnecessary object here or there
- If a endpoint is a list endpoint, and have pageToken and size, then this endpoint need to support pagniation and it need to have the overload syntax to provide concrete type

**Review by orangecoloured — APPROVED** [review]

**Pull request merged**